core/src/main/scala/org/apache/spark/deploy/master/Master.scala

```diff
@@ -561,7 +561,7 @@ private[deploy] class Master(
     apps.filter(_.state == ApplicationState.UNKNOWN).foreach(finishApplication)
 
     // Update the state of recovered apps to RUNNING
-    apps.filter(_.state == ApplicationState.WAITING).foreach(_.state = ApplicationState.RUNNING)
+    apps.filter(_.coresLeft == 0).foreach(_.state = ApplicationState.RUNNING)
```

Comment:
Hmmm...this might not be true. An app with coresLeft > 0 could also be RUNNING as long as it gained an executor from Master:

https://github.com/apache/spark/blob/d9e4cf67c06b2d6daa4cd24b056e33dfb5eb35f5/core/src/main/scala/org/apache/spark/deploy/master/Master.scala#L716-L720

Maybe, using `_.coresGranted > 0` ?

Reply:
Yes you are right. Update it , thank you